core/pbcc_request_retry_timer.c

```diff
@@ -0,0 +1,198 @@
+ * polling and give other processes some CPU time.
+ */
+#define PBCC_POLLING_TIMEOUT 10
+
+
+// ----------------------------------------------
+//                     Types
+// ----------------------------------------------
+
+#if defined      _WIN32
+typedef DWORD    pubnub_thread_t;
+typedef FILETIME pubnub_timespec_t;
+typedef void     pubnub_watcher_t;
+#define thread_handle_field() HANDLE thread_handle;
+#define UNIT_IN_MILLI 1000
+#else
+typedef pthread_t       pubnub_thread_t;
+typedef struct timespec pubnub_timespec_t;
+typedef void*           pubnub_watcher_t;
+#define thread_handle_field()
+#endif
+
+/** Failed requests retry timer definition. */
+struct pbcc_request_retry_timer {
+    /** PubNub context for which timer will restart requests. */
+    pubnub_t* pb;
+#if defined(PUBNUB_CALLBACK_API)
+    /** Timer's run thread. */
+    pubnub_thread_t timer_thread;
+#endif // #if defined(PUBNUB_CALLBACK_API)
+    /** Whether timer is running at this moment or not. */
+    volatile bool running;
+    /** Active timer delay value in milliseconds. */
+    uint16_t delay;
+    /** Shared resources access lock. */
+    pubnub_mutex_t mutw;
+};
+
+
+// ----------------------------------------------
+//              Function prototypes
+// ----------------------------------------------
+
+
+/**
+ * @brief Run request retry timer.
+ *
+ * Launch timer which will await till timeout and try to restart request if
+ * possible.
+ *
+ * @param timer Pointer to the timer which should run.
+ */
+static void* pbcc_request_retry_timer_run_(pbcc_request_retry_timer_t* timer);
+
+
+// ----------------------------------------------
+//                   Functions
+// ----------------------------------------------
+
+pbcc_request_retry_timer_t* pbcc_request_retry_timer_alloc(pubnub_t* pb)
+{
+    PBCC_ALLOCATE_TYPE(timer, pbcc_request_retry_timer_t, true, NULL);
+    pubnub_mutex_init(timer->mutw);
+#if defined(PUBNUB_CALLBACK_API)
+    timer->timer_thread = NULL;
+#endif // #if defined(PUBNUB_CALLBACK_API)
+    timer->running = false;
+    timer->pb      = pb;
+
+    return timer;
+}
+
+void pbcc_request_retry_timer_free(pbcc_request_retry_timer_t** timer)
+{
+    if (NULL == timer || NULL == *timer) { return; }
+
+    pbcc_request_retry_timer_stop(*timer);
+    pubnub_mutex_destroy((*timer)->mutw);
+    free(*timer);
+    *timer = NULL;
+}
+
+void pbcc_request_retry_timer_start(
+    pbcc_request_retry_timer_t* timer,
+    const uint16_t              delay)
+{
+    if (NULL == timer || timer->running) { return; }
+    PUBNUB_ASSERT(pb_valid_ctx_ptr(timer->pb));
+
+    /** There shouldn't be any active timers. */
+    pbcc_request_retry_timer_stop(timer);
+
+    pubnub_mutex_lock(timer->mutw);
+    timer->running = true;
+    timer->delay   = delay;
+    pubnub_mutex_unlock(timer->mutw);
+    pubnub_mutex_lock(timer->pb->monitor);
+    timer->pb->state = PBS_WAIT_RETRY;
+    timer->pb->core.http_retry_count++;
+    pubnub_mutex_unlock(timer->pb->monitor);
```

Comment:
Q: is it okay to lock those locks separetely?
Won't it cause any UB?

Reply:
@Xavrax I don't think it is critical, but I can put one of them inside another after sanitizer check (I hope it is not critical).